File: backend/src/main/scala/com/skewwhiffy/auraltester/scales/Scale.scala
package com.skewwhiffy.auraltester.scales

import com.skewwhiffy.auraltester.notes.AbsoluteNote

import scala.util.chaining.scalaUtilChainingOps

class Scale(
  private val _lowestNote: AbsoluteNote,
  val scaleType: ScaleType,
  val direction: ScaleDirection
) {
  def lowestNote: AbsoluteNote = _lowestNote

  def displayName: String = s"${lowestNote.note.displayString} ${scaleType.displayName}"

  def notes: List[AbsoluteNote] = scaleType
    .intervals(direction)
    .map(lowestNote.apply)
    .pipe(it => direction match {
      case ScaleDirection.ascending => it
      case ScaleDirection.descending => it.reverse
      // TODO: Check what happens with invalid choice
    })

  def abc(key: Key): String = notes.map(it => it.abc(key)).mkString("")
}

File: backend.kotlin/src/main/kotlin/com/skewwhiffy/notation/model/scale/Scale.kt
package com.skewwhiffy.notation.model.scale

import com.skewwhiffy.notation.model.key.Key
import com.skewwhiffy.notation.model.note.AbsoluteNote
import com.skewwhiffy.notation.model.note.NoteSequence

data class Scale(
  val lowestNote: AbsoluteNote,
  private val scaleType: ScaleType,
  val direction: ScaleDirection,
) : NoteSequence {
  val displayName: String = "${lowestNote.note.displayString} ${scaleType.displayName}"

  override val notes: List<AbsoluteNote>
    get() = scaleType
      .intervals(direction)
      .map(lowestNote::apply)
      .let {
        when (direction) {
          ScaleDirection.ASCENDING -> it
          ScaleDirection.DESCENDING -> it.reversed()
        }
      }

  fun abc(key: Key): String = notes.joinToString("") { it.abc(key) }
}